### PR TITLE
pass in text-rendered-webpage instead of outputString

### PR DESCRIPTION
**Pull request by seanmcguire12**
# why
ideally, this improves upon the following issues with `extract`:
- ensure the whole webpage is processed instead of stopping processing too early
- improve overall accuracy of `extract`, ie, for larger extraction tasks, collect more of the correct data and less of the irrelevant/incorrect data 
- provide the model/LLM with a more accurate depiction of what the webpage actually looks like (in text form)

# what changed
Instead of passing HTML to the LLM, we now pass a text-rendered representation of the webpage. Ideally this gives the LLM more context about how the webpage is actually organized, and reduces unnecessary tokens (HTML tags, indices, etc)

# test plan
- create more evals for larger extractions with multiple fields in the schema
- create more evals on complex websites, (ecommerce sites, javascript heavy sites, etc)

### CODE REVIEW

**Comment by navidkpr** [comment]
I rebased the new backend in with your code. Could not push to your branch so had to create a branch here for it: 
[sean/extract_improvement](https://github.com/browserbase/stagehand/tree/sean/extract_improvement)

Could u create a pr from your account from that branch? Or give me access to your fork and I'll push it in for u

File: lib/handlers/extractHandler.ts [hunk 1]
```diff
@@ -138,42 +189,25 @@ export class StagehandExtractHandler {
...
         category: "extraction",
-        message: "got response",
-        auxiliary: {
-          extraction_response: {
-            value: JSON.stringify(extractionResponse),
-            type: "object",
-          },
-        },
+        message: "extraction completed successfully",
+        level: 1,
```
Comment (kamath) [inline lib/handlers/extractHandler.ts:356]: add `extraction_response` under auxiliary for consistency

File: lib/handlers/extractHandler.ts [hunk 2]
```diff
@@ -85,46 +83,99 @@ export class StagehandExtractHandler {
...
+        }
+      }
+    }
+
+    await this.stagehand.page.evaluate((dom) => window.restoreDOM(dom), originalDOM);
+    const formattedText = formatText(textAnnotations);
+
+    // const fs = require("fs");
+    // const formattedTextFilePath = "./formattedText.txt";
+    // fs.writeFileSync(formattedTextFilePath, formattedText);
```
Comment (kamath) [inline lib/handlers/extractHandler.ts]: remove and add to structured logs instead

File: lib/utils.ts [hunk 3]
```diff
@@ -1,10 +1,228 @@
...
 
+export function formatText(textAnnotations: TextAnnotation[]): string {
+  // cluster tokens by line
+  const lineCluster: Map<number, TextAnnotation[]> = new Map();
+  const lineThreshold = 10;
+
+  // sort annotations by their Y-coordinate
+  const sortedAnnotations = textAnnotations.sort((a, b) => a.midpoint.y - b.midpoint.y);
+
+  sortedAnnotations.forEach((annotation) => {
```
Comment (kamath) [inline lib/utils.ts:35]: add some comments here, this is a little hard to parse out of context

File: lib/utils.ts [hunk 4]
```diff
@@ -1,10 +1,228 @@
...
+    const lastY = keys[keys.length - 1];
+
+    if (keys.length > 0 && Math.abs(lineY - lastY) < lineThreshold) {
+      lineCluster.get(lastY)!.push(annotation);
+    } else {
+      lineCluster.set(lineY, [annotation]);
+    }
+  });
+
+  let canvasWidth = 80;
```
Comment (kamath) [inline lib/utils.ts:48]: dumb question, what's a canvas and why are you setting this number here?

File: lib/utils.ts [hunk 5]
```diff
@@ -1,10 +1,228 @@
...
+
+  // calculate canvas width based on the longest line
```
Comment (kamath) [inline lib/utils.ts:50]: what longest line? again, some more general comments here would be super helpful to understand what's going on

File: lib/utils.ts [hunk 6]
```diff
@@ -1,10 +1,228 @@
 import crypto from "crypto";
 import { LogLine } from "../types/log";
+import { TextAnnotation } from "../types/textannotation";
 
 export function generateId(operation: string) {
   return crypto.createHash("sha256").update(operation).digest("hex");
 }
 
+export function formatText(textAnnotations: TextAnnotation[]): string {
```
Comment (kamath) [inline lib/utils.ts:10]: General theme here -- appreciate the comments you do have, but I think a more generalized docstring that outlines the process you're taking and why, and then linking the comments back to steps in said docstring would be super helpful here. Right now, I understand that this works better, but if I had to debug this I'd be super lost -- what is a canvas? What are the widths you're calculating? What are the different lines you're referencing?

File: lib/handlers/extractHandler.ts [hunk 7]
```diff
@@ -85,46 +83,99 @@ export class StagehandExtractHandler {
 
     await this.waitForSettledDom(domSettleTimeoutMs);
     await this.startDomDebug();
```
Comment (kamath) [inline lib/handlers/extractHandler.ts:199]: Note for future: analyze where these are called and also clean up these functions to avoid redundancy

File: lib/handlers/extractHandler.ts [hunk 8]
```diff
@@ -85,46 +83,99 @@ export class StagehandExtractHandler {
...
-        },
-        chunks_total: {
-          value: chunks.length.toString(),
-          type: "integer",
-        },
-      },
+      message: `received output from processAllOfDom. selectorMap has ${Object.keys(selectorMap).length} entries`,
+      level: 1,
     });
+    const PROXIMITY_THRESHOLD = 10;
```
Comment (kamath) [inline lib/handlers/extractHandler.ts]: what is this?

File: lib/handlers/extractHandler.ts [hunk 9]
```diff
@@ -85,46 +83,99 @@ export class StagehandExtractHandler {
...
+    await this.stagehand.page.evaluate(() => window.createTextBoundingBoxes());
+    const pageWidth = await this.stagehand.page.evaluate(() => window.innerWidth);
+    const pageHeight = await this.stagehand.page.evaluate(() => window.innerHeight);
+
+    const seenAnnotations = new Map();
+    const textAnnotations: TextAnnotation[] = [];
+
```
Comment (kamath) [inline lib/handlers/extractHandler.ts:233]: comment: start to process xpaths from selectorMap

File: lib/handlers/extractHandler.ts [hunk 10]
```diff
@@ -85,46 +83,99 @@ export class StagehandExtractHandler {
...
+    for (const xpaths of Object.values(selectorMap)) {
+      const xpath = xpaths[0];
+
+      const boundingBoxes: Array<{
```
Comment (kamath) [inline lib/handlers/extractHandler.ts:241]: why is this an array?

Comment (kamath) [reply lib/handlers/extractHandler.ts:241]: from call: an xpath will take you to an element, and each element will have a list of words, each word with their own bounding box. this is that array

Comment (kamath) [reply lib/handlers/extractHandler.ts:241]: `boundingBoxes` is a list of each word within an xpath element

File: lib/handlers/extractHandler.ts [hunk 11]
```diff
@@ -85,46 +83,99 @@ export class StagehandExtractHandler {
...
+        left: number;
+        top: number;
+        width: number;
+        height: number;
+      }> = await this.stagehand.page.evaluate(
+        (xpath) => window.getElementBoundingBoxes(xpath),
+        xpath
+      );
+
+      for (const box of boundingBoxes) {
```
Comment (kamath) [inline lib/handlers/extractHandler.ts:252]: nit: rename to word or wordbox or something

File: lib/handlers/extractHandler.ts [hunk 12]
```diff
@@ -85,46 +83,99 @@ export class StagehandExtractHandler {
...
+        const text = box.text;
+
+        let annotationsForText = seenAnnotations.get(text);
+        if (!annotationsForText) {
+          annotationsForText = [];
+          seenAnnotations.set(text, annotationsForText);
+        }
+
+        const isDuplicate = annotationsForText.some((annotation) => {
```
Comment (kamath) [inline lib/handlers/extractHandler.ts]: The parent element candidate could have children that are also candidates. We want to prevent this from happening

File: lib/handlers/extractHandler.ts [hunk 13]
```diff
@@ -85,46 +83,99 @@ export class StagehandExtractHandler {
...
+          const dx = annotation.x - (box.left + box.width / 2);
+          const dy = annotation.y - (box.top + box.height / 2);
+          const distance = Math.sqrt(dx * dx + dy * dy);
+          return distance < PROXIMITY_THRESHOLD;
```
Comment (kamath) [inline lib/handlers/extractHandler.ts]: Check for deduping by checking bounding box + same word; this works but doesn't get rid of all of them. Not sure why but there are situations where it'll fail (won't remove duplicates or will still add). 

Thus, add proximity threshold to give you a tolerance for deduping. This is hacky. Is "10" 10px?

File: lib/handlers/extractHandler.ts [hunk 14]
```diff
@@ -85,46 +83,99 @@ export class StagehandExtractHandler {
...
+        height: number;
+      }> = await this.stagehand.page.evaluate(
+        (xpath) => window.getElementBoundingBoxes(xpath),
+        xpath
+      );
+
+      for (const box of boundingBoxes) {
+        const text = box.text;
+
+        let annotationsForText = seenAnnotations.get(text);
```
Comment (kamath) [inline lib/handlers/extractHandler.ts]: can make this const?

File: lib/handlers/extractHandler.ts [hunk 15]
```diff
@@ -85,46 +83,99 @@ export class StagehandExtractHandler {
...
+        if (!annotationsForText) {
```
Comment (kamath) [inline lib/handlers/extractHandler.ts]: if text not in seenAnnotations, set text to empty list (const defined above)

File: lib/handlers/extractHandler.ts [hunk 16]
```diff
@@ -85,46 +83,99 @@ export class StagehandExtractHandler {
...
+
+        if (!isDuplicate) {
+          annotationsForText.push({
+            x: box.left + box.width / 2,
+            y: box.top + box.height / 2,
+          });
+
+          textAnnotations.push({
+            text: box.text,
+            midpoint: {
```
Comment (kamath) [inline lib/handlers/extractHandler.ts]: leave a comment explaining why this is not actually the midpoint

File: lib/handlers/extractHandler.ts [hunk 17]
```diff
@@ -85,46 +83,99 @@ export class StagehandExtractHandler {
...
+              x: box.left,
+              y: box.top + box.height,
+            },
+            midpoint_normalized: {
```
Comment (kamath) [inline lib/handlers/extractHandler.ts]: relative to the width of the page

File: lib/handlers/extractHandler.ts [hunk 18]
```diff
@@ -2,8 +2,52 @@ import { LLMProvider } from "../llm/LLMProvider";
...
 import { z } from "zod";
 import { LogLine } from "../../types/log";
+import { TextAnnotation } from "../../types/textannotation";
 import { extract } from "../inference";
 import { LLMClient } from "../llm/LLMClient";
+import { formatText } from "../utils";
+
+const PROXIMITY_THRESHOLD = 15;
+
+/**
```
Comment (kamath) [inline lib/handlers/extractHandler.ts:12]: amazing work, love this!! smallest nit: can you change the tense of the verbs here to align with the comments below? will help with cmd+F for a specific step.

File: lib/handlers/extractHandler.ts [hunk 19]
```diff
@@ -2,8 +2,52 @@ import { LLMProvider } from "../llm/LLMProvider";
...
+ * The `StagehandExtractHandler` class is responsible for extracting structured data from a webpage by:
+ *
+ * **1. Waiting for the DOM to settle and initializing DOM debugging.**
```
Comment (kamath) [inline lib/handlers/extractHandler.ts]:
```suggestion
 * **1:** Wait for the DOM to settle and start DOM debugging
```

etc.

File: lib/handlers/extractHandler.ts [hunk 20]
```diff
@@ -2,8 +2,52 @@ import { LLMProvider } from "../llm/LLMProvider";
...
 import { LogLine } from "../../types/log";
+import { TextAnnotation } from "../../types/textannotation";
 import { extract } from "../inference";
 import { LLMClient } from "../llm/LLMClient";
+import { formatText } from "../utils";
+
+const PROXIMITY_THRESHOLD = 15;
+
+/**
+ * The `StagehandExtractHandler` class is responsible for extracting structured data from a webpage by:
```
Comment (kamath) [inline lib/handlers/extractHandler.ts]:
```suggestion
 * The `StagehandExtractHandler` class is responsible for extracting structured data from a webpage. Here's what it does at a high level:
```

File: lib/handlers/extractHandler.ts [hunk 21]
```diff
@@ -2,8 +2,52 @@ import { LLMProvider } from "../llm/LLMProvider";
...
+ *
+ * **1. Waiting for the DOM to settle and initializing DOM debugging.**
+ *    - Ensures the page is fully loaded and stable before extraction.
+ *
+ * **2. Storing the original DOM before any mutations.**
```
Comment (kamath) [inline lib/handlers/extractHandler.ts]:
```suggestion
 * **2:** Store the original DOM before any mutations
```

File: lib/handlers/extractHandler.ts [hunk 22]
```diff
@@ -2,8 +2,52 @@ import { LLMProvider } from "../llm/LLMProvider";
...
+ *    - Preserves the initial state of the DOM to restore later.
```
Comment (kamath) [inline lib/handlers/extractHandler.ts:22]: would love like just one more sentence here talking about how this is necessary since you mentioned on call that it's really hard to just add elements and remove and keep the previous DOM. no need to elaborate too much, maybe just that one sentence + a link or something would be 🔥

File: lib/handlers/extractHandler.ts [hunk 23]
```diff
@@ -2,8 +2,52 @@ import { LLMProvider } from "../llm/LLMProvider";
...
+ *
+ * **3. Processing the DOM to generate a selector map of candidate elements.**
```
Comment (kamath) [inline lib/handlers/extractHandler.ts]:
```suggestion
 * **3:** Process the DOM to generate a selector map of candidate elements
```

File: lib/handlers/extractHandler.ts [hunk 24]
```diff
@@ -2,8 +2,52 @@ import { LLMProvider } from "../llm/LLMProvider";
...
+ *    - Identifies potential elements that contain the data to extract.
+ *
+ * **4. Creating text bounding boxes around every word in the webpage.**
```
Comment (kamath) [inline lib/handlers/extractHandler.ts]:
```suggestion
 * **4:** Create text bounding boxes around every word in the webpage
```

File: lib/handlers/extractHandler.ts [hunk 25]
```diff
@@ -2,8 +2,52 @@ import { LLMProvider } from "../llm/LLMProvider";
...
+ *    - Wraps words in spans so that their bounding boxes can be used to
+ *      determine their positions on the text-rendered-webpage.
+ *
+ * **5. Collecting all text annotations (with positions and dimensions) from each of the candidate elements.**
```
Comment (kamath) [inline lib/handlers/extractHandler.ts]:
```suggestion
 * **5:** Collect all text annotations (with positions and dimensions) from the candidate elements
```

File: lib/handlers/extractHandler.ts [hunk 26]
```diff
@@ -2,8 +2,52 @@ import { LLMProvider } from "../llm/LLMProvider";
...
+ *    - Gathers text and positional data for each word.
+ *
+ * **6. Grouping annotations by text and deduplicating them based on proximity.**
```
Comment (kamath) [inline lib/handlers/extractHandler.ts]:
```suggestion
 * **6:** Group annotations by text and deduplicate them based on proximity
```

File: lib/handlers/extractHandler.ts [hunk 27]
```diff
@@ -2,8 +2,52 @@ import { LLMProvider } from "../llm/LLMProvider";
...
+ *    - There is no guarantee that the text annotations are unique (candidate elements can be nested).
+ *    - Thus, we must remove duplicate words that are close to each other on the page.
+ *
+ * **7. Restoring the original DOM after mutations.**
```
Comment (kamath) [inline lib/handlers/extractHandler.ts]:
```suggestion
 * **7:** Restore the original DOM after mutations
```

File: lib/handlers/extractHandler.ts [hunk 28]
```diff
@@ -2,8 +2,52 @@ import { LLMProvider } from "../llm/LLMProvider";
...
+ *    - Returns the DOM to its original state after processing.
+ *
+ * **8. Formatting the deduplicated annotations into a text representation.**
```
Comment (kamath) [inline lib/handlers/extractHandler.ts]: Don't see this one?

File: lib/handlers/extractHandler.ts [hunk 29]
```diff
@@ -2,8 +2,52 @@ import { LLMProvider } from "../llm/LLMProvider";
...
+ *    - Prepares the text data for the extraction process.
+ *
+ * **9. Passing the formatted text to an LLM for extraction according to the given instruction and schema.**
```
Comment (kamath) [inline lib/handlers/extractHandler.ts]:
```suggestion
 * **9:** Pass the formatted text to an LLM for extraction according to the given instruction and schema
```

File: lib/handlers/extractHandler.ts [hunk 30]
```diff
@@ -2,8 +2,52 @@ import { LLMProvider } from "../llm/LLMProvider";
...
+ *    - Uses a language model to extract structured data based on instructions.
+ *
+ * **10. Handling the extraction response and logging the results.**
```
Comment (kamath) [inline lib/handlers/extractHandler.ts]:
```suggestion
 * **10:** Handle the extraction response and log the results
```

File: lib/handlers/extractHandler.ts [hunk 31]
```diff
@@ -83,50 +123,135 @@ export class StagehandExtractHandler {
       },
     });
 
+    // **Step 1:** Wait for the DOM to settle and start DOM debugging
```
Comment (kamath) [inline lib/handlers/extractHandler.ts]:
```suggestion
    // **1:** Wait for the DOM to settle and start DOM debugging
```

File: lib/handlers/extractHandler.ts [hunk 32]
```diff
@@ -83,50 +123,135 @@ export class StagehandExtractHandler {
...
     await this.waitForSettledDom(domSettleTimeoutMs);
     await this.startDomDebug();
-    const { outputString, chunk, chunks } = await this.stagehand.page.evaluate(
-      (chunksSeen?: number[]) => window.processDom(chunksSeen ?? []),
-      chunksSeen,
-    );
+
+    // **Step 2:** Store the original DOM before any mutations
```
Comment (kamath) [inline lib/handlers/extractHandler.ts]:
```suggestion
    // **2:** Store the original DOM before any mutations
```

File: lib/handlers/extractHandler.ts [hunk 33]
```diff
@@ -83,50 +123,135 @@ export class StagehandExtractHandler {
...
+    // we need to store the original DOM here because calling createTextBoundingBoxes()
+    // will mutate the DOM by adding spans around every word
+    const originalDOM = await this.stagehand.page.evaluate(() => window.storeDOM());
+
+    // **Step 3:** Process the DOM to generate a selector map of candidate elements
```
Comment (kamath) [inline lib/handlers/extractHandler.ts]:
```suggestion
    // **3:** Process the DOM to generate a selector map of candidate elements
```

File: lib/handlers/extractHandler.ts [hunk 34]
```diff
@@ -83,50 +123,135 @@ export class StagehandExtractHandler {
...
-        chunks_total: {
-          value: chunks.length.toString(),
-          type: "integer",
-        },
-      },
+      message: `received output from processAllOfDom. selectorMap has ${Object.keys(selectorMap).length} entries`,
+      level: 1,
     });
 
+    // **Step 4:** Create text bounding boxes around every word in the webpage
```
Comment (kamath) [inline lib/handlers/extractHandler.ts]:
```suggestion
    // **4:** Create text bounding boxes around every word in the webpage
```

File: lib/handlers/extractHandler.ts [hunk 35]
```diff
@@ -83,50 +123,135 @@ export class StagehandExtractHandler {
...
+    // calling createTextBoundingBoxes() will create a span around every word on the
+    // webpage. The bounding boxes of these spans will be used to determine their
+    // positions in the text rendered webpage
+    await this.stagehand.page.evaluate(() => window.createTextBoundingBoxes());
+    const pageWidth = await this.stagehand.page.evaluate(() => window.innerWidth);
+    const pageHeight = await this.stagehand.page.evaluate(() => window.innerHeight);
+
+    // **Step 5:** Collect all text annotations (with positions and dimensions) from the candidate elements
```
Comment (kamath) [inline lib/handlers/extractHandler.ts]:
```suggestion
    // **5:** Collect all text annotations (with positions and dimensions) from the candidate elements
```

File: lib/handlers/extractHandler.ts [hunk 36]
```diff
@@ -83,50 +123,135 @@ export class StagehandExtractHandler {
...
+          bottom_left,
+          bottom_left_normalized,
+          width: box.width,
+          height: box.height,
+        };
+        allAnnotations.push(annotation);
+      }
+    }
+
+    // **Step 6:** Group annotations by text and deduplicate them based on proximity
```
Comment (kamath) [inline lib/handlers/extractHandler.ts]:
```suggestion
    // **6:** Group annotations by text and deduplicate them based on proximity
```

File: lib/handlers/extractHandler.ts [hunk 37]
```diff
@@ -83,50 +123,135 @@ export class StagehandExtractHandler {
...
+          return distance < PROXIMITY_THRESHOLD;
+        });
+
+        if (!isDuplicate) {
+          deduplicatedTextAnnotations.push(annotation);
+        }
+      }
+    }
+
+    // **Step 7:** Restore the original DOM after mutations
```
Comment (kamath) [inline lib/handlers/extractHandler.ts]:
```suggestion
    // **7:** Restore the original DOM after mutations
```

File: lib/handlers/extractHandler.ts [hunk 38]
```diff
@@ -83,50 +123,135 @@ export class StagehandExtractHandler {
...
+    await this.stagehand.page.evaluate((dom) => window.restoreDOM(dom), originalDOM);
+    const formattedText = formatText(deduplicatedTextAnnotations);
+
+    // **Step 9:** Pass the formatted text to an LLM for extraction according to the given instruction and schema
```
Comment (kamath) [inline lib/handlers/extractHandler.ts]:
```suggestion
    // **9:** Pass the formatted text to an LLM for extraction according to the given instruction and schema
```

File: lib/handlers/extractHandler.ts [hunk 39]
```diff
@@ -83,50 +123,135 @@ export class StagehandExtractHandler {
...
 
     const {
-      metadata: { progress: newProgress, completed },
+      metadata: { completed },
       ...output
     } = extractionResponse;
+
     await this.cleanupDomDebug();
 
+    // **Step 10:** Handle the extraction response and log the results
```
Comment (kamath) [inline lib/handlers/extractHandler.ts]:
```suggestion
    // **10:** Handle the extraction response and log the results
```

**Review by kamath — COMMENTED** [review]
Great work, this is phenomenal! Minor commit suggestions for easier cmd+F'ing. Also I think you're missing step 8.
